device/core/src/client_diagnostic.ts

```diff
@@ -0,0 +1,206 @@
+    if (this.shouldAddDiagnosticInfo()) {
+      let diagnostics: Diagnostics = new Diagnostics(this.generateEightRandomCharacters(), this.getCurrentTimeUtc());
+      message.diagnostics = diagnostics;
+    }
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.onDesiredTwinUpdate
+   * @description       callback when desired twin received.
+   *
+   * @param {Twin}   twin           The twin object
+   * @param {any}    samplingRate    The sampling rate
+   */
+  onDesiredTwinUpdate(twin: Twin, samplingRate: any): void {
+    /* Codes_SRS_NODE_DEVICE_DIAGNOSTICCLIENT_01_007: [The onDesiredTwinUpdate function shall be a callback when client receive a desired twin update.] */
+    let errMsg = '';
+    /* Codes_SRS_NODE_DEVICE_DIAGNOSTICCLIENT_01_009: [The onDesiredTwinUpdate function shall set the sampling percentage.] */
+    // We would like to handle all value types for samplingRate here, if the type is not number we will report in the twin.
+    // That's why the samplingRate type is any.
+    if (typeof samplingRate === 'number') {
+      try {
+        this.setDiagSamplingPercentage(samplingRate);
+      } catch (e) {
+        errMsg += `Value of ${DiagnosticClient.DIAGNOSTIC_TWIN_KEY_SAMPLING_PERCENTAGE} is invalid:${e.message} `;
+      }
+    } else if (samplingRate === null) {
+      this.setDiagSamplingPercentage(0);
+    } else {
+      errMsg += `Value of ${DiagnosticClient.DIAGNOSTIC_TWIN_KEY_SAMPLING_PERCENTAGE} is invalid:${samplingRate}(check if it is number) `;
+    }
+    /* Codes_SRS_NODE_DEVICE_DIAGNOSTICCLIENT_01_010: [The onDesiredTwinUpdate function shall send a reported twin to give feedback to diagnostic update.] */
+    twin.properties.reported.update({
+      [DiagnosticClient.DIAGNOSTIC_TWIN_KEY_SAMPLING_PERCENTAGE]: this.diagSamplingPercentage,
+      [DiagnosticClient.DIAGNOSTIC_TWIN_KEY_ERROR_MSG]: errMsg
+    }, (err) => {
+      if (err) {
+        debug('Update reported twin for diagnostic failed:' + err.message);
+      }
+    });
+  }
+
+  // Get a character from 0-9a-zA-Z
+  private getDiagnosticIdChar(value: number): string {
+    if (value <= 9) {
+      // 48 is char code of '0'
+      return String.fromCharCode(48 + value);
+    } else if (value <= 9 + 26) {
+      // 65 is char code of 'A'
+      return String.fromCharCode(65 + value - 10);
+    } else {
+      // 97 is char code of 'a'
+      return String.fromCharCode(97 + value - 36);
+    }
+  }
+
+  private generateEightRandomCharacters(): string {
+    // This function shall generate 8 random chars, each is from 0-9a-z.
+    let result: string = '';
+    for (let i = 0; i < 8; i++) {
+      let randomNum: number = Math.floor(Math.random() * DiagnosticClient.DIAGNOSTIC_ID_CHARACTER_BASE);
+      result += this.getDiagnosticIdChar(randomNum);
+    }
+    return result.toString();
+  }
```

Comment:
suggested reading: [https://stackoverflow.com/questions/9407892/how-to-generate-random-sha1-hash-to-use-as-id-in-node-js](https://stackoverflow.com/questions/9407892/how-to-generate-random-sha1-hash-to-use-as-id-in-node-js)